pkg/hubble/relay/server/server_test.go

```diff
@@ -0,0 +1,308 @@
+	)
+	require.NoError(t, err)
+	return pp
+}
+
+var endpoints map[string]*testutils.FakeEndpointInfo
+
+func genStringSlice(prefix string, n int) (ret []string) {
+	for i := 0; i < n; i++ {
+		ret = append(ret, fmt.Sprintf("%s-%d", prefix, i))
+	}
+	return
+}
+
+func init() {
+	endpoints = make(map[string]*testutils.FakeEndpointInfo, 254)
+	namespaces := genStringSlice("namespace", 10)
+	pods := genStringSlice("very-long-pod-name", 50)
+	labels := genStringSlice("quite-long-label", 60)
+	for i := 0; i < 254; i++ {
+		ip := fmt.Sprintf("192.0.2.%d", i)
+		endpoints[ip] = &testutils.FakeEndpointInfo{
+			ID:           uint64(i),
+			IPv4:         net.ParseIP(ip),
+			PodNamespace: namespaces[i%len(namespaces)],
+			PodName:      pods[i%len(pods)],
+			Labels:       labels[i%(len(labels)-10) : i%(len(labels)-10)+10],
+		}
+	}
+}
+
+func newHubbleObserver(t testing.TB, nodeName string, numFlows int) *observer.LocalObserverServer {
+	queueSize := numFlows
+
+	pp := noopParser(t)
+	s, err := observer.NewLocalServer(pp, log,
+		observeroption.WithMaxFlows(container.Capacity65535),
+		observeroption.WithMonitorBuffer(queueSize),
+	)
+	require.NoError(t, err)
+
+	m := s.GetEventsChannel()
+
+	for i := 0; i < numFlows; i++ {
+		tn := monitor.TraceNotifyV0{Type: byte(monitorAPI.MessageTypeTrace)}
+		srcIP := fmt.Sprintf("192.0.2.%d", i%len(endpoints))
+		dstIP := fmt.Sprintf("192.0.2.%d", (i+10)%len(endpoints))
+		srcMAC, _ := net.ParseMAC("00:00:5e:00:53:01")
+		dstMAC, _ := net.ParseMAC("00:00:5e:00:53:02")
+		data := testutils.MustCreateL3L4Payload(tn,
+			&layers.Ethernet{
+				SrcMAC:       srcMAC,
+				DstMAC:       dstMAC,
+				EthernetType: layers.EthernetTypeIPv4,
+			},
+			&layers.IPv4{
+				SrcIP: endpoints[srcIP].IPv4,
+				DstIP: endpoints[dstIP].IPv4,
+			},
+			&layers.TCP{
+				SrcPort: 123,
+				DstPort: 456,
+				ACK:     true,
+				PSH:     i%4 == 0,
+			})
+		event := &observerTypes.MonitorEvent{
+			Timestamp: time.Unix(int64(i+1), 0),
+			NodeName:  nodeName,
+			Payload: &observerTypes.PerfEvent{
+				Data: data,
+				CPU:  0,
+			},
+		}
+		m <- event
+	}
+
+	return s
+}
+
+func newHubblePeer(t testing.TB, ctx context.Context, address string, hubbleObserver *observer.LocalObserverServer) {
+	options := []serveroption.Option{
+		serveroption.WithInsecure(),
+		serveroption.WithUnixSocketListener(address),
+		serveroption.WithObserverService(hubbleObserver),
+	}
+
+	srv, err := server.NewServer(log, options...)
+	require.NoError(t, err)
+
+	go func() {
+		if err := srv.Serve(); err != nil {
+			t.Log(err)
+			t.Fail()
+		}
+	}()
+
+	go func() {
+		<-ctx.Done()
+		close(hubbleObserver.GetEventsChannel())
+		<-hubbleObserver.GetStopped()
+		srv.Stop()
+	}()
+}
+
+func benchmarkRelayGetFlows(b *testing.B, withFieldMask bool) {
+	tmp, err := os.MkdirTemp("", "hubble")
+	require.NoError(b, err)
+	defer os.RemoveAll(tmp)
+	root := "unix://" + filepath.Join(tmp, "peer-")
+	ctx := context.Background()
+	numFlows := b.N
```

Comment:
Yes, it would be better. However, I wanted to test the entire stack as most allocations happen during proto transmission in this case. The real problem (as stated in FIXME) is scaling with the number of peers which is a workaround to issues I had with sending more flows than fit in a ring buffer.

I am happy to apply any suggestions to fix that, but currently I don't have a good idea how to implement this.